src/sources/mod.rs

```diff
@@ -157,25 +164,28 @@ pub trait EventSource {
     /// [`Poll::unregister`](crate::Poll#method.unregister) method.
     fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()>;
 
-    /// Notification that a polling session is going to start
+    /// Whether this source needs to be sent the [`EventSource::before_sleep`]
+    /// and [`EventSource::before_handle_events`] notifications. These are opt-in because
+    /// they require more expensive checks, and almost all sources will not need these notifications
+    const NEEDS_EXTRA_LIFECYCLE_EVENTS: bool = false;
+    /// Notification that a single `poll` is about to begin.
+    /// If this returns Ok(Some), polling will shortcircuit, and
+    /// your event handler will be called with the returned Token and Readiness
```

Comment:
The current implementation does still invoke poll with a timeout of 0 even if the `before_sleep()` method returned `Ok(Some(...))`, which means it is possible that poll generates another legit event. We need to document what happens in this case.

What the current implementation does is:
- `before_handle_events()` is invoked if a real event was generated, independently of what `before_sleep()` returned
- If a real event is generated in addition to the one from `before_sleep()`, then `dispatch_events()` is invoked twice. 

Imo this is the correct thing to do (to handle composite sources with multiple tokens), but this needs to be documented clearly.